adding integration with lambdatest (#713)
codemagic-ci-cd/codemagic-docs: content/integrations/lambdatest-integration.md

## Create a LambdaTest account

Registering with LambdaTest is required in order to be able to get the username and access token. You can sign up for free [here](https://www.lambdatest.com/).

## Configuring Real Time and App Automation with Codemagic

**LambdaTest** offers two testing environments, **Real Time** and **App Automation**. Applications can be submitted to both testing environments through Codemagic using a cURL request. In order to configure them correctly, you will need two pieces of data: **username** and **access token**. They can be found in the **LambdaTest UI** with your account. 

## Real Time and App Automation

You can test your **.ipa** and **.apk** directly on real devices by submitting them to the **LambdaTest** environment via a cURL request:

```
-Scripts:
    - name: Submitting app to LambdaTest:
    script: |
        curl --location --request POST 'https://manual-api.lambdatest.com/app/upload/realDevice' --header 'Authorization: Basic $LAMBDATEST' --form 'name="lambda1"' --form 'appFile=@"app/build/outputs/apk/release/app-release.apk"'
```

> **$LAMBDATEST** environment variable is a custom name and can be named to anything preferred. More info can be found [here](https://docs.codemagic.io/variables/environment-variable-groups/) about how to set up variable groups with Codemagic. The environment variable is the base64 encoded form of your username and access token. They need to be encoded with the following format [here](https://mixedanalytics.com/knowledge-base/api-connector-encode-credentials-to-base-64/):

Reviewer: Perhaps just change the order of information here. After you say which information is needed to integrate Lambdatest and Codemagic, I would continue with encoding the username and token and saving the encoded string to Codemagic as an environment variable. So the steps would be as follow:
1) Encode and save the username and access token into an env var
2) Run the scripts in yaml referencing that env var

Author: Changes were made.

> **$LAMBDATEST** environment variable is a custom name and can be named to anything preferred. More info can ber found [here](https://docs.codemagic.io/variables/environment-variable-groups/) about how to set up variable groups with Codemagic. The environment variable is the base64 encoded form of your username and access token. They need to be encoded with the following format [here](https://mixedanalytics.com/knowledge-base/api-connector-encode-credentials-to-base-64/):

Reviewer: This section here is confusing for me. In the previous section, you mentioned that 2 environment variables are needed but here you are mentioning just one.
Also I'm not sure why base64-encoding is needed, is that so as to not expose the values during the build?

Author: I might need to change "two env vars" to "two pieces of data" because Lambdatest requires it be base64 encoded together like username:accesstoken

Author: Yes, most probably it is because of not exposing them.

---
title: LambdaTest integration
description: How to integrate your workflows with LambdaTest using codemagic.yaml
weight: 6
---

> **LambdaTest** is a cloud-based cross browser testing tool that allows you to perform cross browser testing for mobile applications. You can perform both manual and automated cross browser testing using LambdaTest. More information can be found here. 

Reviewer 2: You have mentioned "cross browser" three times in this sentence. Surely we are testing mobile applications, not web apps on multiple browsers??

Author: Done!

## Configuring Real Time and App Automation with Codemagic

**LambdaTest** offers two testing environments, **Real Time** and **App Automation**. Applications can be submitted to both testing environments through Codemagic using a cURL request. In order to configure them correctly, you will need two pieces of data: **username** and **access token**. They can be found in the **LambdaTest UI** with your account. 

## Real Time and App Automation

You can test your **.ipa** and **.apk** directly on real devices by submitting them to the **LambdaTest** environment via a cURL request:

```
-Scripts:
    - name: Submitting app to LambdaTest:
    script: |
        curl --location --request POST 'https://manual-api.lambdatest.com/app/upload/realDevice' --header 'Authorization: Basic $LAMBDATEST' --form 'name="lambda1"' --form 'appFile=@"app/build/outputs/apk/release/app-release.apk"'
```

Base64 encoding the username and token [here](https://mixedanalytics.com/knowledge-base/api-connector-encode-credentials-to-base-64/) with the **username:accessToken** format is needed, then save the encoded string to Codemagic as an environment variable. So the steps would be as follow:

> 1.Encode and save the username and access token into an env var.
> 2.Run the scripts in yaml referencing that env var.

Reviewer 2: ```suggestion
1. Encode and save the username and access token into an environment variable.
2. Run the scripts in YAML referencing that environment variable.
```

Author: done

> **$LAMBDATEST** environment variable is a custom name and can be named to anything preferred. More info can be found [here](https://docs.codemagic.io/variables/environment-variable-groups/) about how to set up variable groups with Codemagic.

Reviewer 2: ```suggestion
**$LAMBDATEST** environment variable is a custom name and can be named to anything preferred. More info can be found [here](https://docs.codemagic.io/variables/environment-variable-groups/) about how to set up environment variable groups with Codemagic.
```

Author: done

> **$LAMBDATEST** environment variable is a custom name and can be named to anything preferred. More info can be found [here](https://docs.codemagic.io/variables/environment-variable-groups/) about how to set up environment variable groups with Codemagic.

Reviewer: The information about username and token and how to save and reference them in Codemagic should come in the previous section before you give the actual script that uses them.

Author: Would not it be better that we give the cURL and then explain it ? Otherwise, it could be confusing

Reviewer: I guess both approaches have their merits, so up to you. :)

Author: I myself usually find it very easy when I see the end product / result and then its explanation underneath. So, I'd suggest keeping it this way ))

As soon as your **.ipa** and **.apk** are successfully built, they will appear in the **LambdatTest UI** under **Real Device => Real Time**. Any preferred devices can be selected for testing with **Real Time**. 

In order to see your tests being uploaded to the **App Automation**, tests need to be included in your project. As soon as tests are detected, they will be automatically uploaded to the **App Automation** section and all the results can be viewed there. However, in order to enable it, some capabilities must be injected into your project tests:

```
DesiredCapabilities capabilities = new DesiredCapabilities(); capabilities.setCapability("platformName", "Android"); capabilities.setCapability("deviceName", "Google Pixel 3"); capabilities.setCapability("isRealMobile", true); capabilities.setCapability("platformVersion","10"); capabilities.setCapability("app","lt://APP100202151634649275590734"); capabilities.setCapability("deviceOrientation", "PORTRAIT"); capabilities.setCapability("console",true); capabilities.setCapability("network",true); capabilities.setCapability("visual",true);

> ```

Reviewer 2: Which file are these set in?

Author: test files

As soon as your **.ipa** and **.apk** are successfully built, they will appear in the **LambdatTest UI** under **Real Device => Real Time**. Any preferred devices can be selected for testing with **Real Time**. 

In order to see your tests being uploaded to the **App Automation**, tests need to be included in your project. As soon as tests are detected, they will be automatically uploaded to the **App Automation** section and all the results can be viewed there. However, in order to enable it, some capabilities must be injected into your project's test scripts:

**For Android**:

```    
    DesiredCapabilities capabilities = new DesiredCapabilities();
    capabilities.setCapability("platformName", "Android");
    capabilities.setCapability("deviceName", "Google Pixel 3");
    capabilities.setCapability("isRealMobile", true);
    capabilities.setCapability("platformVersion","10");
    capabilities.setCapability("app","lt://APP100202151634649275590734");
    capabilities.setCapability("deviceOrientation", "PORTRAIT");
    capabilities.setCapability("console",true);
    capabilities.setCapability("network",true);
    capabilities.setCapability("visual",true);

> ```

Reviewer 2: Please mention which file in the project this is required in.

**For iOS**:

```
    DesiredCapabilities capabilities = new DesiredCapabilities();
    capabilities.setCapability("platformName", "iOS");
    capabilities.setCapability("deviceName", "iPhone 10");
    capabilities.setCapability("isRealMobile", true);
    capabilities.setCapability("platformVersion","10");
    capabilities.setCapability("app","lt://APP100202151634649275590734");
    capabilities.setCapability("deviceOrientation", "PORTRAIT");
    capabilities.setCapability("console",true);
    capabilities.setCapability("network",true);
    capabilities.setCapability("visual",true);

> ```

Reviewer 2: Mention the file where this is set.

Author: Hmm, this should be set in a test script and that test script file could be located differently, I guess. Should we just leave it as a test script rather than showing an exact file path ?

These capabitlies will allow **LambdaTest** to detect which platform you want to you execute your test scripts with. In these capabilities, the main part is **app URL** which is generated in the response of the cURL request:

```
{"app_id":"APP10020171164383758036593","name":"lambda1","type":"android","app_url":"lt://APP10020171444643838005433352"}
```

## Sample project

> A sample project that shows how to configure LambdaTes integration is available [here](https://github.com/codemagic-ci-cd/codemagic-sample-projects/tree/main/integrations/lambdatest_integration_demo_project/lambdatest_integration_demo_project).

Reviewer 2: ```suggestion
A sample project that shows how to configure LambdaTest integration is available [here](https://github.com/codemagic-ci-cd/codemagic-sample-projects/tree/main/integrations/lambdatest_integration_demo_project/lambdatest_integration_demo_project).
```
